File: libs/ai-kernel/src/models/knowledge_base.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use serde_json::Value;
use sqlx::{FromRow, types::Json};
use uuid::Uuid;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct KnowledgeChunk {
    pub id: String,
    pub position: i32,
    pub text: String,
    pub token_count: i32,
    #[serde(default)]
    pub embedding: Vec<f32>,
    #[serde(default)]
    pub metadata: Value,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct KnowledgeDocument {
    pub id: Uuid,
    pub knowledge_base_id: Uuid,
    pub title: String,
    pub content: String,
    pub source_uri: Option<String>,
    pub metadata: Value,
    pub status: String,
    pub chunk_count: i32,
    pub chunks: Vec<KnowledgeChunk>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct KnowledgeBase {
    pub id: Uuid,
    pub name: String,
    pub description: String,
    pub status: String,
    pub embedding_provider: String,
    pub chunking_strategy: String,
    pub tags: Vec<String>,
    pub document_count: i64,
    pub chunk_count: i64,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct KnowledgeSearchResult {
    pub knowledge_base_id: Uuid,
    pub document_id: Uuid,
    pub document_title: String,
    pub chunk_id: String,
    pub score: f32,
    pub excerpt: String,
    pub source_uri: Option<String>,
    pub metadata: Value,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ListKnowledgeBasesResponse {
    pub data: Vec<KnowledgeBase>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ListKnowledgeDocumentsResponse {
    pub data: Vec<KnowledgeDocument>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CreateKnowledgeBaseRequest {
    pub name: String,
    #[serde(default)]
    pub description: String,
    #[serde(default = "default_knowledge_status")]
    pub status: String,
    #[serde(default = "default_embedding_provider")]
    pub embedding_provider: String,
    #[serde(default = "default_chunking_strategy")]
    pub chunking_strategy: String,
    #[serde(default)]
    pub tags: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct UpdateKnowledgeBaseRequest {
    pub name: Option<String>,
    pub description: Option<String>,
    pub status: Option<String>,
    pub embedding_provider: Option<String>,
    pub chunking_strategy: Option<String>,
    pub tags: Option<Vec<String>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CreateKnowledgeDocumentRequest {
    pub title: String,
    pub content: String,
    #[serde(default)]
    pub source_uri: Option<String>,
    #[serde(default)]
    pub metadata: Value,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SearchKnowledgeBaseRequest {
    pub query: String,
    #[serde(default = "default_search_top_k")]
    pub top_k: usize,
    #[serde(default = "default_min_score")]
    pub min_score: f32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SearchKnowledgeBaseResponse {
    pub knowledge_base_id: Uuid,
    pub query: String,
    pub results: Vec<KnowledgeSearchResult>,
    pub retrieved_at: DateTime<Utc>,
}

#[derive(Debug, Clone, FromRow)]
pub(crate) struct KnowledgeBaseRow {
    pub id: Uuid,
    pub name: String,
    pub description: String,
    pub status: String,
    pub embedding_provider: String,
    pub chunking_strategy: String,
    pub tags: Json<Vec<String>>,
    pub document_count: i64,
    pub chunk_count: i64,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, FromRow)]
pub(crate) struct KnowledgeDocumentRow {
    pub id: Uuid,
    pub knowledge_base_id: Uuid,
    pub title: String,
    pub content: String,
    pub source_uri: Option<String>,
    pub metadata: Json<Value>,
    pub status: String,
    pub chunk_count: i32,
    pub chunks: Json<Vec<KnowledgeChunk>>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

impl From<KnowledgeBaseRow> for KnowledgeBase {
    fn from(value: KnowledgeBaseRow) -> Self {
        Self {
            id: value.id,
            name: value.name,
            description: value.description,
            status: value.status,
            embedding_provider: value.embedding_provider,
            chunking_strategy: value.chunking_strategy,
            tags: value.tags.0,
            document_count: value.document_count,
            chunk_count: value.chunk_count,
            created_at: value.created_at,
            updated_at: value.updated_at,
        }
    }
}

impl From<KnowledgeDocumentRow> for KnowledgeDocument {
    fn from(value: KnowledgeDocumentRow) -> Self {
        Self {
            id: value.id,
            knowledge_base_id: value.knowledge_base_id,
            title: value.title,
            content: value.content,
            source_uri: value.source_uri,
            metadata: value.metadata.0,
            status: value.status,
            chunk_count: value.chunk_count,
            chunks: value.chunks.0,
            created_at: value.created_at,
            updated_at: value.updated_at,
        }
    }
}

fn default_knowledge_status() -> String {
    "active".to_string()
}

fn default_embedding_provider() -> String {
    "deterministic-hash".to_string()
}

fn default_chunking_strategy() -> String {
    "balanced".to_string()
}

fn default_search_top_k() -> usize {
    5
}

fn default_min_score() -> f32 {
    0.55
}

File: libs/ai-kernel-go/models/knowledge_base.go
package models

import (
	"encoding/json"
	"time"

	"github.com/google/uuid"
)

// KnowledgeChunk is a single retrievable chunk of a document.
type KnowledgeChunk struct {
	ID         string          `json:"id"`
	Position   int32           `json:"position"`
	Text       string          `json:"text"`
	TokenCount int32           `json:"token_count"`
	Embedding  []float32       `json:"embedding"`
	Metadata   json.RawMessage `json:"metadata"`
}

// KnowledgeDocument carries the full document body + its chunks.
type KnowledgeDocument struct {
	ID              uuid.UUID        `json:"id"`
	KnowledgeBaseID uuid.UUID        `json:"knowledge_base_id"`
	Title           string           `json:"title"`
	Content         string           `json:"content"`
	SourceURI       *string          `json:"source_uri"`
	Metadata        json.RawMessage  `json:"metadata"`
	Status          string           `json:"status"`
	ChunkCount      int32            `json:"chunk_count"`
	Chunks          []KnowledgeChunk `json:"chunks"`
	CreatedAt       time.Time        `json:"created_at"`
	UpdatedAt       time.Time        `json:"updated_at"`
}

// KnowledgeBase is the catalog row for one KB.
type KnowledgeBase struct {
	ID                uuid.UUID `json:"id"`
	Name              string    `json:"name"`
	Description       string    `json:"description"`
	Status            string    `json:"status"`
	EmbeddingProvider string    `json:"embedding_provider"`
	ChunkingStrategy  string    `json:"chunking_strategy"`
	Tags              []string  `json:"tags"`
	DocumentCount     int64     `json:"document_count"`
	ChunkCount        int64     `json:"chunk_count"`
	CreatedAt         time.Time `json:"created_at"`
	UpdatedAt         time.Time `json:"updated_at"`
}

// KnowledgeSearchResult is one citation hit returned by RAG queries.
type KnowledgeSearchResult struct {
	KnowledgeBaseID uuid.UUID       `json:"knowledge_base_id"`
	DocumentID      uuid.UUID       `json:"document_id"`
	DocumentTitle   string          `json:"document_title"`
	ChunkID         string          `json:"chunk_id"`
	Score           float32         `json:"score"`
	Excerpt         string          `json:"excerpt"`
	SourceURI       *string         `json:"source_uri"`
	Metadata        json.RawMessage `json:"metadata"`
}

type ListKnowledgeBasesResponse struct {
	Data []KnowledgeBase `json:"data"`
}

type ListKnowledgeDocumentsResponse struct {
	Data []KnowledgeDocument `json:"data"`
}

// CreateKnowledgeBaseRequest defaults: status="active",
// embedding_provider="deterministic-hash", chunking_strategy="balanced".
type CreateKnowledgeBaseRequest struct {
	Name              string   `json:"name"`
	Description       *string  `json:"description"`
	Status            *string  `json:"status"`
	EmbeddingProvider *string  `json:"embedding_provider"`
	ChunkingStrategy  *string  `json:"chunking_strategy"`
	Tags              []string `json:"tags"`
}

type UpdateKnowledgeBaseRequest struct {
	Name              *string   `json:"name"`
	Description       *string   `json:"description"`
	Status            *string   `json:"status"`
	EmbeddingProvider *string   `json:"embedding_provider"`
	ChunkingStrategy  *string   `json:"chunking_strategy"`
	Tags              *[]string `json:"tags"`
}

type CreateKnowledgeDocumentRequest struct {
	Title     string          `json:"title"`
	Content   string          `json:"content"`
	SourceURI *string         `json:"source_uri"`
	Metadata  json.RawMessage `json:"metadata"`
}

// SearchKnowledgeBaseRequest defaults: top_k=5, min_score=0.55.
type SearchKnowledgeBaseRequest struct {
	Query    string  `json:"query"`
	TopK     uint32  `json:"top_k"`
	MinScore float32 `json:"min_score"`
}

func (r *SearchKnowledgeBaseRequest) UnmarshalJSON(data []byte) error {
	type alias SearchKnowledgeBaseRequest
	aux := struct {
		TopK     *uint32  `json:"top_k"`
		MinScore *float32 `json:"min_score"`
		*alias
	}{
		alias: (*alias)(r),
	}
	if err := json.Unmarshal(data, &aux); err != nil {
		return err
	}
	if aux.TopK == nil {
		r.TopK = DefaultSearchTopK
	} else {
		r.TopK = *aux.TopK
	}
	if aux.MinScore == nil {
		r.MinScore = DefaultSearchMinScore
	} else {
		r.MinScore = *aux.MinScore
	}
	return nil
}

type SearchKnowledgeBaseResponse struct {
	KnowledgeBaseID uuid.UUID               `json:"knowledge_base_id"`
	Query           string                  `json:"query"`
	Results         []KnowledgeSearchResult `json:"results"`
	RetrievedAt     time.Time               `json:"retrieved_at"`
	SearchProvider  string                  `json:"search_provider,omitempty"`
	SearchMode      string                  `json:"search_mode,omitempty"`
}

const (
	DefaultKnowledgeStatus           = "active"
	DefaultEmbeddingProvider         = "deterministic-hash"
	DefaultChunkingStrategy          = "balanced"
	DefaultSearchTopK        uint32  = 5
	DefaultSearchMinScore    float32 = 0.55
)

func (r *CreateKnowledgeBaseRequest) UnmarshalJSON(data []byte) error {
	type alias CreateKnowledgeBaseRequest
	if err := json.Unmarshal(data, (*alias)(r)); err != nil {
		return err
	}
	if r.Description == nil {
		r.Description = ptrOf("")
	}
	if r.Status == nil {
		r.Status = ptrOf(DefaultKnowledgeStatus)
	}
	if r.EmbeddingProvider == nil {
		r.EmbeddingProvider = ptrOf(DefaultEmbeddingProvider)
	}
	if r.ChunkingStrategy == nil {
		r.ChunkingStrategy = ptrOf(DefaultChunkingStrategy)
	}
	if r.Tags == nil {
		r.Tags = []string{}
	}
	return nil
}

func (r *CreateKnowledgeDocumentRequest) UnmarshalJSON(data []byte) error {
	type alias CreateKnowledgeDocumentRequest
	if err := json.Unmarshal(data, (*alias)(r)); err != nil {
		return err
	}
	r.Metadata = defaultRawMessage(r.Metadata, emptyJSONObject())
	return nil
}
